### PR TITLE
Async copy over ssh

### PR DESCRIPTION
**Pull request by paskino**
Adds a class to handle async copies of file over the SSH link.

Example usage at https://github.com/paskino/CIL-work/blob/3f8e8f88602bf3801eff73eea102f61bcc3cadde/RemoteViewer/TwoLinkedViewersCentralWidgetCompare.py#L155

### CODE REVIEW

**Comment by paskino** [comment]
This only retrieves from an SSH server. It should be bidirectional.

File: brem/AsyncCopyOverSSH.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            os.chdir(localdir)
+            if direction == 'from':
+                action = 'get'
+                if status_callback is not None:
+                    status_callback.emit("{} file {}".format(action, filename))
+                a.get_file("{}".format(filename))
+            else:
+                action = 'put'
+                # dest_fname   = kwargs.get('dest_fname', None)
+                dest_fname = posixpath.join(self.remotedir, filename)
```
Comment (paskino) [inline brem/AsyncCopyOverSSH.py:142]: this should make sure of what is the remote OS and change path accordingly.

File: brem/AsyncCopyOverSSH.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    def SetCopyToRemote(self):
+        self.direction = 'to'
+
+    def SetCopyFromRemote(self):
+        self.direction = 'from'
+    
+    def SetRemoteOS(self, value):
+        allowed_os = ['Windows', 'POSIX']
+        if value in allowed_os:
+            self.remote_os = value
```
Comment (paskino) [inline brem/AsyncCopyOverSSH.py:42]: the reason to set `remote_os` is to use `posixpath` or `ntpath`. 
Probably this class could have a default `POSIX` and this method should just configure the `remotepath` that can be used later:

```python

if value == 'POSIX':
    self.remotepath = posixpath
elif value == 'Windows':
    self.remotepath = ntpath
else:
    raise ValueError('Expected value in {}. Got {}'.format(['POSIX','Windows'], value))

```

File: brem/AsyncCopyOverSSH.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                                  filename = self.filename,
+                                  localdir=localdir,
+                                  direction = self.direction,
+                                  host=host, 
+                                  username=username, 
+                                  port=port, 
+                                  private_key=private_key, 
+                                  logfile=self.logfile, 
+                                  update_delay=10,
+                                  remote_os=remote_os
```
Comment (paskino) [inline brem/AsyncCopyOverSSH.py:71]: if `self.remotepath` is available this parameter can be removed

File: brem/AsyncCopyOverSSH.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    def copy_worker(self, **kwargs):
+        # retrieve the appropriate parameters from the kwargs
+        host         = kwargs.get('host', None)
+        username     = kwargs.get('username', None)
+        port         = kwargs.get('port', None)
+        private_key  = kwargs.get('private_key', None)
+        logfile      = kwargs.get('logfile', None)
+        update_delay = kwargs.get('update_delay', None)
+        direction    = kwargs.get('direction', None)
+        remote_os    = kwargs.get('remote_os', None)
```
Comment (paskino) [inline brem/AsyncCopyOverSSH.py:91]: as above

File: brem/AsyncCopyOverSSH.py (hunk 1 above)
Comment (paskino) [reply brem/AsyncCopyOverSSH.py:142]: use `self.remotepath`

File: brem/AsyncCopyOverSSH.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+import paramiko
+from brem import BasicRemoteExecutionManager
+import os, ntpath, posixpath
+import socket
+
+class RemoteAsyncCopyOverSSHSignals(QtCore.QObject):
+    status = QtCore.Signal(tuple)
+    job_id = QtCore.Signal(int)
+class AsyncCopyOverSSH(object):
+    '''Class to handle async copy of files over SSH'''
```
Comment (paskino) [inline brem/AsyncCopyOverSSH.py:12]: expand on the docstring. Add `remote_os` in the parameter for the creator as [_often-changed_](https://github.com/TomographicImaging/CIL/discussions/1064#discussioncomment-1657282).

**Review by paskino — COMMENTED** [review]
Add docstrings to most of the methods that don't have anything, especially the ones that accept a parameter.

Add string constants `POSIX` and `Windows`

File: brem/AsyncCopyOverSSH.py (hunk 3 above)
Comment (paskino) [reply brem/AsyncCopyOverSSH.py:107]: remote_os is required by `BaseRemoteExecutionManager`

File: brem/__init__.py [hunk 6]
```diff
@@ -29,6 +29,7 @@
     from .brem import BasicRemoteExecutionManager
     from .brem import RemoteRunControl
     from .brem import RemoteRunControlSignals
+    from .AsyncCopyOverSSH import AsyncCopyOverSSH, RemoteAsyncCopyOverSSHSignals
     
 
 except ImportError:
```
Comment (lauramurgatroyd) [inline brem/__init__.py:35]: If we change to
```
except ImportError as e:
    raise ImportError('error importing brem: ', e)
```

Then we will see the details of the error in future

File: brem/__init__.py [hunk 7]
```diff
@@ -29,6 +29,7 @@
     from .brem import BasicRemoteExecutionManager
```
Comment (lauramurgatroyd) [inline brem/__init__.py:29]: need to add:
`from .brem import Windows, POSIX`
above the other imports 

This should fix the failing gh-action

File: brem/__init__.py (hunk 6 above)
Comment (paskino) [reply brem/__init__.py:35]: What would it change?

**Review by lauramurgatroyd — APPROVED** [review]
I tested using scarf (using copy_from_ssh and also the viewer example) and it is working well!

**Pull request merged**